koku/masu/api/additional_context.py

```diff
@@ -0,0 +1,53 @@
+#
+# Copyright 2023 Red Hat Inc.
+# SPDX-License-Identifier: Apache-2.0
+#
+"""View for modifying the additional context field of a schema."""
+from django.views.decorators.cache import never_cache
+from rest_framework import status
+from rest_framework.decorators import api_view
+from rest_framework.decorators import permission_classes
+from rest_framework.decorators import renderer_classes
+from rest_framework.permissions import AllowAny
+from rest_framework.response import Response
+from rest_framework.settings import api_settings
+
+from masu.database.provider_db_accessor import ProviderDBAccessor
+
+
+@never_cache
+@api_view(http_method_names=["GET", "POST"])
+@permission_classes((AllowAny,))
+@renderer_classes(tuple(api_settings.DEFAULT_RENDERER_CLASSES))
+def additional_context(request):
+    """Returns or modifies the additional context field."""
+    params = request.query_params
+
+    if not params.get("schema"):
+        errmsg = "Parameter missing. Required: schema"
+        return Response({"Error": errmsg}, status=status.HTTP_400_BAD_REQUEST)
+
+    if not params.get("provider_uuid"):
+        errmsg = "Parameter missing. Required: provider_uuid"
+        return Response({"Error": errmsg}, status=status.HTTP_400_BAD_REQUEST)
+
+    with ProviderDBAccessor(params.get("provider_uuid")) as provider_accessor:
+        context = provider_accessor.get_additional_context()
+        errmsg = None
+        if request.method == "POST":
+            data = request.data
+            for key, value in data.items():
+                if key in provider_accessor.provider.ADDITIONAL_CONTEXT_KEYS:
+                    if value in [True, False]:
+                        context[key] = value
+                    else:
+                        errmsg = f"Invalid value supplied: key: {key}, value: {value}."
+                elif key == "remove_key":
+                    if value in context.keys():
+                        del context[value]
+                else:
+                    errmsg = f"Invalid key supplied: {key}"
+                if errmsg:
+                    return Response({"Error": errmsg}, status=status.HTTP_400_BAD_REQUEST)
```

Comment:
This logic is a bit confusing to reason about. I think it is easier validate the keys/values first and immediately return the error:
```suggestion
                if key not in ["aws_list_account_aliases", "crawl_hierarchy", "remove_key"]:
                    return Response({"Error": f"Invalid key supplied: {key}"}, status=status.HTTP_400_BAD_REQUEST)
                if key == "remove_key" and value in context.keys():
                        del context[value]
                elif key in ["aws_list_account_aliases", "crawl_hierarchy"]:
                    if not isinstance(value, bool):
                        return Response({"Error": f"Invalid value supplied: key: {key}, value: {value}."}, status=status.HTTP_400_BAD_REQUEST)
                    context[key] = value
```

I'm also curious, why was `ADDITIONAL_CONTEXT_KEYS` made an attribute of the Provider class? Is that attribute used anywhere else? It feels like a weird place to put that when it could just be a constant in this file instead.

Reply:
I threw it in there since it related to a field in the Provider's model. I am okay with moving to to a constant in this file instead though.